### ISSUE TITLE
Warn on typed array pollution

### ISSUE THREAD

**Issue by donquixote** [vimeo/psalm#6930]
Currently if a local variable is declared to have a type like `@var string $x` or `@var array{name: string} $info`, this type is only valid until it is broken by an assignment, e.g. `$x = (int) $x;` or `$info['name'] = 5;`.
https://psalm.dev/r/932892536d
https://psalm.dev/r/719d01722e
If later we want to use the variable in a context where the original type is required, we see an error - but it is not reported where the variable loses its original type.

This behavior can be desirable, if the initial `@var` statement only describes the _current_ state of the variable at the time, not any possible future state.

However, in other cases it would be desirable to define the full range of types a variable can have throughout its life cycle.

A possible way to have both and avoid BC breaks would be a new annotation for that purpose.

```php
/** @return string */
function f() {
    /** @var-permanent string $x */
    $x = '5';
    $x = (int) $x;  // Psalm error, because variable changes type.
    return $x;  // Psalm error, because variable no longer matches expected type for return.
}
```

Perhaps I missed something, and such an annotation already exists?
Otherwise, I think it would be a good idea to add it. Perhaps find a better name.

**Comment by donquixote** [comment]
A variable type change can be especially mysterious if nested array values are assigned at different positions within a loop.

**Comment by orklah** [comment]
I can understand the appeal this could have, but it's still pretty limited. You don't actually want to add an inline docblock on each and every variable in your code, and you can't really targets variables that could causes bugs such as this feature would prevent neither.

This means it would need to be a global flag (or like, only on parameter, or only on properties...)

Aside from that, what does "changing type" means? 
- Is `'a'` a change if the variable was a `string`?
- Is `''` a change if the variable was `'a'`?
- Is `never` a change if variable was `''`?
- Is `int|string` a change if variable was `numeric`?
- Is `bool` a change if the variable was `scalar`

I think it's very impractical, but if you manage to propose a consistent PR, we can discuss this again

**Comment by donquixote** [comment]
I am changing the issue title, because it was misleading.
Initially I was only interested in arrays, but then I noticed that this applies on a broader scope.

> Aside from that, what does "changing type" means?

Any deviation from type declared through `@var`.
By "declared type" we could either use existing `@var` docs on local vars - using this would be a BC break or change of behavior. Or we could introduce a new annotation. Or we could make it a setting in psalm.xml.

> Is 'a' a change if the variable was a string?
What matters is whether it had a `/** @var string $x */` doc. If so, then `$x = 'a'` is not a violation.

> Is '' a change if the variable was 'a'?
If the variable was declared as `/** @var 'a' $x */` then `$x = ''` is a violation.
If the variable was declared as `/** @var string $x */`, then assigning first `'a'` and then `''` is both fine.

> Is never a change if variable was ''?
> Is int|string a change if variable was numeric?
> Is bool a change if the variable was scalar
How can we make an assignment that sets the variable to `never` or `int|string` or `bool`?
Perhaps by assigning the value from another var..

```php
/** @param int|string $arg */
function f($arg) {
  /** @var numeric $x */
  $x = 5;
  $x = $arg;  // Violation of declared type for $x, because $arg could be a non-numeric string.
}
```
Psalm already has logic and rules to deal with these cases.

----

It seems that currently, property docs are "prescriptive", whereas local var docs are "descriptive", meaning they describe the current value, without making any restriction on future changes.
All I am asking here is to allow "descriptive" type declarations for local vars.

**Comment by donquixote** [comment]
Hmm what's confusing?
Does my post deviate too much from the initial post?

**Comment by donquixote** [comment]
Oh, I see, quotes were messed up. Fixed that.

**Comment by orklah** [comment]
Do you actually have a use case where:
- you know in advance a variable is sensitive enough to make an effort flagging it with an inline docblock type
- a change in said code is likely to happen such as to change the type to a wider type
- Psalm would not already catch such change as erroneous
- you actually know the type of the variable at one point in the code (because the docblock type would override the inferred type) but you have a reason why you don't want to check this type at a later point in the code (by using assertions for example)
?

As a side note, there is an ongoing project on PHPStan side that will probably change the meaning of `@var` in the future in order to be less "prescriptive". I'm waiting to see exactly what comes out of it, but we'll probably follow in some way because `@var` is not a sane way to deal with types

Anyway, I won't work on this suggestion myself, so feel free to propose something in a PR if you want to make this a reality

**Comment by donquixote** [comment]
> Do you actually have a use case where: [...]

In general, this is most useful for variables that are used to "collect data", and that are changed multiple times within an algorithm.



### Use case: Structural array types

My planned use case:
I am writing a parser where I want to use an array to collect modifiers, doc comments, a type (e.g. "class" vs "interface" vs "trait" etc), attributes, and possibly other things. The array would have a type declaration using a `@psalm-type` typedef for a structured array.
I am already using similar array typedefs in test classes, to describe values I am reading from and comparing to yml.

I considered using arrays instead of value objects, because:
- Arrays, unlike objects, are passed "by value", which makes them, in a way, "immutable" with regard to side effects on other places.
- The arrays are used for internal data transport, not part of a public API.
- Arrays are a natural return value from yml or csv, whereas to get a value object we need to have some kind of conversion. (this applies only to the use case where I am using it for test classes)
- Performance: (I could be wrong on these) (the main purpose of the parser is to be faster than nikic/php-parser, so...)
  - Arrays are probably faster, if in the average case most of the properties will be empty. Value objects always need to support the empty properties.
  - Resetting an array to `[]` is probably cheaper than cloning or creating an empty value object.
  - Writing array values is probably faster than calling setters or "withers" (immutable setters which clone the object) on value objects. Perhaps directly assigning a public property is faster, but going through the setter is probably slower.

### Use case: Nested array types
```php
/** @param array<class-string, list<C>> $objects */
function f($objects) {
  /** @var array<class-string, list<C>> $itemss */
  $itemss = [];
  // Sorry, this loop is a bit redundant, but should show the idea.
  foreach ($objectss as $objects) {
    foreach ($objects as $object) {
      $itemss[get_class($object)][] = $object;  // Good.
    }
  }
  $itemss[C::class]['x'] = new C();  // Bad, because $itemss[*] must not have string keys.
}

### Use case: Other types (string, object etc)
I think it is generally a good practice to not change the type of a variable over its life cycle. Instead, a new variable should be used. If you want, this could be seen as a kind of "invariant" that makes it easier to verify the code.

E.g.
```php
function bad() {
  $names = ['arcturo', 'beatrix'];
  $names[] = 'pedalfly';
  $names = implode(', ', $names);
}
function good() {
  $names = ['arcturo', 'beatrix'];
  $names[] = 'pedalfly';  // Ok, because type does not change.
  $names_by_comma = implode(', ', $names);  // New var, because new type.
}
```
For longer functions where the value is modified multiple types throughout a loop, a prescriptive type declaration can enforce that a variable remains e.g. a string.

The goal is to detect not immediately visible type changes e.g. when assigning a return value from a function that can return null, or when passing the variable by reference to a function that could change the type. 

### Use case: MixedAssignment
In other issues we were discussing why there is "MixedAssignment" warning if a variable was declared as "mixed" and then arbitrary values are assigned to it.
Having a prescriptive `/** @var mixed */` would explicitly allow any value to be assigned to that variable.

**Comment by donquixote** [comment]
Btw the proposal is very much comparable to typed languages (Java, C++) where local variables have a declared type that won't change. I think this should be the goal, or not?

> Anyway, I won't work on this suggestion myself, so feel free to propose something in a PR if you want to make this a reality

I will see about that when I find the time.
